docs/designs/CodeFlare-SDK-design-doc.md

```diff
@@ -0,0 +1,141 @@
+│   ├── guided_demos/
+│   └── additional_demos/
+├── src/
+│   └── codeflare_sdk/
+│       ├── common/              # Self-contained modules shared across SDK components.
+│       │   ├── kubernetes_cluster/ # APIs for interacting with K8s clusters (e.g., authentication).
+│       │   │   ├── __init__.py
+│       │   │   ├── file.py
+│       │   │   └── test_test.py  # Unit tests contained in the same module.
+│       │   ├── kueue/           # API interactions related to Kueue job queuing.
+│       │   ├── widgets/         # Reusable UI components for visualization.
+│       │   └── utils/           # General utility functions and helpers used across project.
+│       ├── ray/                 # Framework-specific APIs for Ray clusters.
+│       │   ├── cluster/         # Cluster management utilities for Ray.
+│       │   ├── client/          # Ray client interaction modules.
+│       │   └── appwrapper/
+├── tests/                       # End-to-end and upgrade testing modules.
+│   ├── e2e/
+│   └── upgrade/
+├── poetry.lock                  # Project configuration and dependency management (Poetry).
+└── pyproject.toml
+
+```
+
+### Client side best practices:
+1. The SDK should perform client-side validation of user inputs before sending them to the API server, to the extent possible. These validations should be limited to client-side checks, and the SDK is not responsible for performing validations that require server-side data. The SDK does not need to maintain the state of server-side objects. Examples of client-side validations include verifying user inputs, validating YAML schemas, and checking resource requests/limits based on available data.
+2. Any call made to the server should have a client instance passed through it. By default, the SDK should use the same client for all operations unless otherwise specified. For example, the client used to list clusters should also be used to fetch cluster details. The codebase should be designed so that these operations are either methods of the same client (e.g., a k8s client, RayJob client, TrainingClient) or the client instance is passed as an argument to the methods.
+3. Errors returned from the API server, to the extent possible, should be wrapped and handled explicitly.
+
+## Test Coverage:
+The project should aim to maintain reasonable test coverage to ensure code quality. Wherever possible, new lines of code should be accompanied by corresponding test cases. Reviewers are encouraged to exercise discretion when merging pull requests (PRs) that result in a drop in test coverage percentage compared to the current state. To aid in tracking coverage, external tools like Codecov can be integrated with the repository, allowing for easier monitoring and ensuring that the codebase maintains a robust level of test coverage over time.
+
+## Release:
+A new version of CodeFlare SDK will be released once every three weeks.
```

Comment:
That seems pretty frequently?

Reply:
Yeah, that's the process we have been following for entire stack. A release at the end of each sprint.